### DISCUSSION TITLE
Mullvad report : VPN leaks DNS traffic outside the tunnel

### DISCUSSION THREAD

**Discussion by shuvashish76** [celzero/rethink-app#1537] [General]
> Findings
Identified scenarios where the Android OS can leak DNS traffic:
> - If a VPN is active without any DNS server configured.
> - For a short period of time while a VPN app is re-configuring the tunnel or is being force stopped/crashes.

Source: https://mullvad.net/en/blog/dns-traffic-can-leak-outside-the-vpn-tunnel-on-android
Android issue tracker: https://issuetracker.google.com/issues/337961996
Invisible Pro issue: https://github.com/Gedsh/InviZible/issues/262

What's the stand of RDNS on this issue?

Previously mullvad reported similar issue for ["connectivity check trafic"](https://mullvad.net/en/blog/android-leaks-connectivity-check-traffic). _(Muntashir's [CPC](https://f-droid.org/packages/io.github.muntashirakon.captiveportalcontroller/) app work for this purpose)_

**Comment by ignoramous** [comment]
- see: #1167 

re: claims in the article:

> "If a VPN is active without any DNS server configured."

Rethink never leaves DNS servers unconfigured. They're *always* set to `10.111.222.3` (v4) or `fd66:f83a:c650::3` (v6).

https://github.com/celzero/rethink-app/blob/6fc4c28d02cb3b84066b7e0dc432096c4a494ad6/app/src/main/java/com/celzero/bravedns/service/BraveVPNService.kt#L2823-L2831

> "For a short period of time while a VPN app is re-configuring the tunnel or is being force stopped/crashes."

Rethink attempts "seamless handover" (that's been available in Android since Pie, I believe) when possible. "Seamless handover" is Android terminology for avoiding packet leaks for the brief time the tunnel is being re-created.

Rethink may crash, but we're hoping to eliminate those as much as possible: #1455
